src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -0,0 +1,111 @@
+getInterpolatedStubRateFromFpml fr1 fr2 p cal convention = do
+  let
+    tenor1 = fromSome fr1.indexTenor
+    tenor2 = fromSome fr2.indexTenor
+    s = p.adjustedStartDate
+    p1 = adjustDate cal convention (addPeriod s tenor1)
+    p2 = adjustDate cal convention (addPeriod s tenor2)
+    tn = subDate p.adjustedEndDate s
+    t1 = subDate p1 s
+    t2 = subDate p2 s
+    d1 = tn - t1
+    d2 = t2 - tn
+    w1 = intToDecimal d2 / intToDecimal (d1 + d2)
+    w2 = 1.0 - w1
+
+  -- In the ISDA paper, the following expression is used for linear interpolation:
+  --Some (Observe (fr1.floatingRateIndex) + (Observe (fr2.floatingRateIndex) - Observe (fr1.floatingRateIndex)) * Const (intToDecimal (tn - t1) / intToDecimal (t2 - t1)))
+
+  -- This expression can be simplified to:
+  Some $ Const w1 * Observe fr1.floatingRateIndex + Const w2 * Observe fr2.floatingRateIndex
+
+-- | Get the floating stub rate to be used for a stub period.
+getStubRateFloatingFromFpml : (DA.Internal.Record.HasField "adjustedEndDate" r1 Date, DA.Internal.Record.HasField "adjustedStartDate" r1 Date, DA.Internal.Record.HasField "floatingRateIndex" r3 o, DA.Internal.Record.HasField "indexTenor" r3 (Optional Period)) => [r3] -> r1 -> HolidayCalendarData -> BusinessDayConventionEnum -> Optional (Observation t Decimal o)
+getStubRateFloatingFromFpml floatingRates p cal convention = do
+  case floatingRates of
+    [rate] -> getSingleStubRateFromFpml rate
+    [rate1, rate2] -> getInterpolatedStubRateFromFpml rate1 rate2 p cal convention
+    other -> error "only one or two floating stub rates supported"
+
+-- | Get the stub rate to be used for a stub period.
+-- Currently, three main options from the FpML schema are supported:
+-- 1. A fix stubRate.
+-- 2. One or two floating rates for the stub.
+-- 3. No specific stub rate defined -> use the same rate as is used for regular periods.
+getStubRateFromFpml : (DA.Internal.Record.HasField "adjustedEndDate" r1 Date, DA.Internal.Record.HasField "adjustedStartDate" r1 Date, DA.Internal.Record.HasField "finalStub" r2 (Optional r3), DA.Internal.Record.HasField "floatingRate" r3 (Optional [r4]), DA.Internal.Record.HasField "floatingRateIndex" r4 o, DA.Internal.Record.HasField "indexTenor" r4 (Optional Period), DA.Internal.Record.HasField "initialStub" r2 (Optional r3), DA.Internal.Record.HasField "stubRate" r3 (Optional Decimal)) => r2 -> Bool -> r1 -> HolidayCalendarData -> BusinessDayConventionEnum -> Optional (Observation t Decimal o)
+getStubRateFromFpml sc initialStub p cal convention = do
+  let
+    s = if initialStub then fromSome sc.initialStub else fromSome sc.finalStub
+    stubRate = s.stubRate
+  case stubRate of
+    Some stubRate -> if isSome s.floatingRate then error "both stubRate and floatingRate provided" else Some (Const (fromSome s.stubRate))
```

Comment:
```suggestion
    Some stubRate -> if isSome s.floatingRate then error "both stubRate and floatingRate provided" else Const <$> s.stubRate
```

i think this should work, to avoid unwrapping and re-wrapping the optional